tests/test_optimize_shot_adapative.py

```diff
@@ -0,0 +1,417 @@
+
+        opt.s[1] = opt.s[1] // 5  # all array elements have a shot budget of 2
+        opt.s[1][0] = 7    # set the shot budget of the zeroth element to 7
+
+        grad, grad_variance = opt.compute_grad(circuit, args, {})
+        assert len(grad) == 2
+        assert len(grad_variance) == 2
+
+        # check zeroth element of arg 0
+        assert np.allclose(grad[0][0, 0], np.mean(single_shot_grads[0][:8, 0, 0]))
+        assert np.allclose(grad_variance[0][0, 0], np.var(single_shot_grads[0][:8, 0, 0], ddof=1))
+
+        # check other elements of arg 0
+        assert np.allclose(grad[0][0, 1], np.mean(single_shot_grads[0][:5, 0, 1]))
+        assert np.allclose(grad_variance[0][0, 1], np.var(single_shot_grads[0][:5, 0, 1], ddof=1))
+
+        # check zeroth element of arg 1
+        assert np.allclose(grad[1][0], np.mean(single_shot_grads[1][:7, 0]))
+        assert np.allclose(grad_variance[1][0], np.var(single_shot_grads[1][:7, 0], ddof=1))
+
+        # check other elements of arg 1
+        assert np.allclose(grad[1][1], np.mean(single_shot_grads[1][:2, 1]))
+        assert np.allclose(grad_variance[1][1], np.var(single_shot_grads[1][:2, 1], ddof=1))
+
+
+class TestWeightedRandomSampling:
+    """Tests for weighted random Hamiltonian term sampling"""
+
+    def test_wrs_expval_cost(self, mocker):
+        """Checks that cost functions that are expval costs can
+        make use of weighted random sampling"""
+        coeffs = [0.2, 0.1]
+        dev = qml.device("default.qubit", wires=2, analytic=False)
+        H = qml.Hamiltonian(coeffs, [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliZ(1)])
+
+        expval_cost = qml.ExpvalCost(qml.templates.StronglyEntanglingLayers, H, dev)
+        weights = qml.init.strong_ent_layers_normal(n_layers=3, n_wires=2)
+
+        opt = qml.ShotAdaptiveOptimizer(min_shots=10)
+        spy = mocker.spy(opt, "weighted_random_sampling")
+
+        new_weights = opt.step(expval_cost, weights)
+        spy.assert_called_once()
+
+        grads = opt.weighted_random_sampling(expval_cost.qnodes, coeffs, 10, [0], weights)
+        assert len(grads) == 1
+        assert grads[0].shape == (10, *weights.shape)
+
+    def test_wrs_disabled(self, mocker):
+        """Checks that cost functions that are expval costs can
+        disable use of weighted random sampling"""
+        coeffs = [0.2, 0.1]
+        dev = qml.device("default.qubit", wires=2, analytic=False)
+        H = qml.Hamiltonian(coeffs, [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliZ(1)])
+
+        expval_cost = qml.ExpvalCost(qml.templates.StronglyEntanglingLayers, H, dev)
+        weights = qml.init.strong_ent_layers_normal(n_layers=3, n_wires=2)
+
+        opt = qml.ShotAdaptiveOptimizer(min_shots=10, weighted_random_sampling=False)
+        spy = mocker.spy(opt, "weighted_random_sampling")
+
+        new_weights = opt.step(expval_cost, weights)
+        spy.assert_not_called()

```

Comment:
Would it make sense to add an integration test that just uses the optimiser to lower a simple cost (for which we know it can do so)?

Reply:
I was worried it would be difficult to test because of the high degree of stochasticity (both in the gradients, terms that are measured, and how the shots are budgeted). But it would definitely benefit from such a test, would appreciate any thoughts here 🤔

Reply:
I never found a good solution, even in the other optimiser tests :) 

Brute force would just be that after a few steps the cost goes down. This sounds very black-box, but ultimately this is how I always check if code works! If we have enough steps, the stochasticity should not be a problem. And if the example has a simple enough cost function (i.e. qubit rotation) then this should be possible? 

Actually, could be a nice super high-level integration test for all optimisers. If they cannot optimise a trivial case, they are in some sense not working.

Reply:
This is a nice idea - perhaps I will add a basic qubit rotation test

Reply:
Have added two integration tests in ef564a8